design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,278 @@
+
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Supported Versions](#supported-versions)
+- [Production Readiness](#production-readiness)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+This checklist contains actions which must be completed before a PR implementing this design can be merged.
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
+- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]
+
+## Summary
+
+cert-manager has the ability to set the owner reference field in generated Secret resources. The option is global, and takes the form of the flag `--enable-certificate-owner-ref` set in the cert-manager controller Deployment resource.
+
+Let us take an example of Certificate resource:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: cert-1
+  namespace: ns-1
+  uid: 1e0adf8
+spec:
+  secretName: cert-1
+```
+
+When `--enable-certificate-owner-ref` is passed to the cert-manager controller, when issuing the X.509 certificate, cert-manager will create a Secret resource that looks like this:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cert-1
+  namespace: ns-1
+  ownerReferences:
+    - controller: true
+      blockOwnerDeletion: false
+      uid: 1e0adf8
+      name: cert-1
+      kind: Certificate
+      apiVersion: cert-manager.io/v1
+data:
+  tls.crt: "..."
+  tls.key: "..."
+  ca.crt: "..."
+```
+
+The proposition is to add a new field `certificateOwnerRef` to the Certificate resource:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+spec:
+  secretName: cert-1
+  cleanupPolicy: [Delete|Never] # ✨
+```
+
+The new field `cleanupPolicy` has three possible values:
+
+1. When "empty", the behavior will default to not creating an owner reference on the Secret resource, unless `--enable-certificate-owner-ref` is passed.
```

Comment:
Does this mean we don't default this field to the value of the flag during admission, and instead the controller infers the behaviour based on the flag value? Perhaps having empty automatically default to something like "Inherit" or "Default" would be clearer from a user perspective (when running a 'get -o yaml', they'll see the value of this field and be able to infer that the behaviour is based on the flag configuration).

Additionally, if a user changes the cleanupPolicy (either explicitly configured on the Certificate OR a change in the flag value) *after* issuance, will the secret reconciliation controller add (or conversely, remove) the owner reference automatically?

Reply:
If user changes cleanup policy controller do not change existing secrets. Owner reference is applied on the next certificate issuance.

Reply:
> Perhaps having empty automatically default to something like "Inherit" or "Default" would be clearer from a user perspective

I agree, the "empty" value is often avoided in Kubernetes APIs. Let's go with `Inherit` then.

> Will the secret reconciliation controller add (or conversely, remove) the owner reference automatically?

After looking at [`SecretOwnerReferenceManagedFieldMismatch`](https://github.com/cert-manager/cert-manager/blob/f058c815a379afc36e9abff21454fc3423e40f78/internal/controller/certificates/policies/checks.go#L474), it seems like adding the flag `--enable-certificate-owner-ref` when you already have existing Secret resource will add the owner reference.